Chore refactor text input

Karim94 opened the pull request:
#### Relevant info and issue/PR links 
https://app.clubhouse.io/peerio/story/1723/mobile-uniform-styling-for-inputs-signup-login-bordered-inputs
There is a validation bug which was found: "Attempting to login to a deleted or suspended account only gave 'Invalid account key' error instead of 'error_accountDeleted' or 'error_accountSuspended'" Please investigate it.

#### Testing instructions  
Test that validation behaves correctly and as expected by users.
Test toggling between secret mode. Cursor also has to maintain its position.
Test "always dirty" mode, which means that the textbox will always show "error" until the entered text is correct.

----
### Repository owner

- [ ] Was tested and can be merged.
- [ ] PR name follows conventional changelog format.

[PR guideline](https://github.com/PeerioTechnologies/peerio-icebear/blob/dev/docs/CONTRIBUTING.md)

seavan commented on app/components/shared/styled-text-input.js:
```diff
@@ -0,0 +1,145 @@
+    @observable valid = true;
+    @observable showSecret = false;
+
+    onChangeText = (text) => {
```
use @action.bound in components instead of arrow

seavan commented on app/components/shared/styled-text-input.js:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    onBlur = () => {
```
use @action.bound in components instead of arrow

seavan commented on app/components/shared/styled-text-input.js:
```diff
@@ -0,0 +1,145 @@
+        if (this.props.onBlur) this.props.onBlur();
+    }
+
+    onFocus = () => {
```
use @action.bound in components instead of arrow

seavan commented on app/components/shared/styled-text-input.js:
```diff
@@ -0,0 +1,145 @@
+                        onChangeText={this.onChangeText}
+                        onBlur={this.onBlur}
+                        onFocus={this.onFocus}
+                        ref={ref => { this._ref = ref; }}
```
move arrow function to @action.bound member

seavan commented on app/components/shared/styled-text-input.js:
```diff
@@ -0,0 +1,145 @@
+    onChangeText = (text) => {
+        this.props.state.value = text;
+        // Validate if it matches, do NOT invalidate if it doesn't
+        if (this.props.validation(this.props.state.value)) {
```
validation can and will be async (promise).
also, there is a case of "dirty" and non-dirty fields. e.g. when you open signup screen, there are no error messages although the required fields are not yet filled. that's because we consider them "not-dirty".

to account for that, the logic should be as following:

The "dirty" mode should be accounted for in textbox itself, unless textbox has `alwaysDirty` prop set.

No validation should be run before user changes the text, and no validation errors should be displayed.

The validity state is:
not known (validation pending, hasn't been run, no errors are shown)
valid (no errors are shown, all good)
invalid (errors are shown)

When user changes text, validation state immediately changes to `not known`, and validation callback is called.

After validation callback returns, we set the validation state to result of it.

seavan commented:
@Karim94 don't forget to finish it! :)

seavan commented on app/components/mocks/mock-styled-text-input.js:
```diff
@@ -0,0 +1,44 @@
+export default class MockStyleTextInput extends Component {
+    validation1 = async (value) => {
+        const expression = '123';
+        return value === expression ? 1 : 2; // 1 === Valid, 2 === INVALID, because JS enums
```
@Karim94 you should either use:

constants from StyledTextInput (export them with prototype) (this is optimal)

OR true/false/undefined triad.

don't use arbitrary 1, 2 etc.

Karim94 commented:
@seavan Animation is done. "Enum" is done. Not sure how to blur when clicking outside without adding onPress to the outside container that makes `uistate.focusedTextBox = null`.

Karim94 commented:
Animation has been fixed. Textboxes now get blurred when any area outside the textbox is clicked (only in mocks since it requires a scrollview outside the textboxes, but the mocks provide an example for implementation).

seavan commented:
couple of comments:

animation uses JS instead of native. to test if it's ok in our case, @suprko @alexandra-marin please test on the devices (Android, iOS) that focusing/unfocusing inputs is fast and smooth. Simulators are not a good way to test that.

suprko commented:
Looks alright on my very low-end phone. Animation is not the smoothest but it responds quickly.

seavan commented:
![validation-bug](https://user-images.githubusercontent.com/750359/38472986-6554b90c-3b88-11e8-88da-ea58cfa01fdc.gif)

it seems that there's a race condition on validation.

when entering two symbols at a quick rate, the validation end result is validator "Username not found", while it should be format validator.

Karim94 commented on app/components/login/login-clean.js:
```diff
@@ -53,25 +69,33 @@ export default class LoginClean extends LoginWizardPage {
-                                hint={t('title_username')} />
-                            <TextBox key="usernamePassword"
+                            />
+                            {/* TODO need a way to show wrong AK error. This was
```
@seavan please take a look at this when reviewing this PR

seavan replied:
add a method "setCustomError" to StyledTextInput.
calling setCustomError would set validation state to invalid and display that error.

it would be called on unsuccessfull login this.usernameInputRef.setCustomError('login_Error');

text change should reset it.

seavan commented:
about `isValid`.

instead of `isValid` in loginState or signupState, add a method get isValid() to StyledTextInput.

on each wizard page, check if next is available by adding:

get isNextEnabled() {
    return this.usernameRef.isValid && this.passwordRef.isValid
}

etc.

seavan commented:
![image](https://user-images.githubusercontent.com/750359/38669915-c19ab0a6-3e47-11e8-9aea-160c7e46fb0a.png)
it is used here, for example.

> Attempting to login to a deleted or suspended account only gave 'Invalid account key' error instead of 'error_accountDeleted' or 'error_accountSuspended'

that looks like a bug.

seavan commented:
@Karim94 

you can leave it as a note and I will fix it when reviewing it.

if everything is ready to be checked and tested (signup+login), let me know.

Karim94 commented:
@seavan I noted the bug in the description. Please proceed to review and testing.

seavan commented:
@Karim94 please resolve conflicts

@alexandra-marin please test validation on login and signup with new textboxes on iOS, when conflicts are resolved. thanks.

seavan commented:
@alexandra-marin waiting for your test feedback, thanks

Karim94 commented:
@seavan @ebilg 
Summary of this PR's state:
Alexandra has tested this and left her comments. Samvel justified one of them and I corrected the other one.
Then I had a design review with Kauro and fixed all her comments. (No logic has been changed, only UI)
@ebilg this PR is ready for testing.

ebilg commented:
can be merged
